src/d3a/models/myco_matcher/external_matcher.py

```diff
@@ -1,140 +1,179 @@
+        self.myco_ext_conn.publish_json(channel, response_data)
 
     def match_recommendations(self, message):
         """Receive trade recommendations and match them in the relevant market.
 
         Matching in bulk, any pair that fails validation will cancel the operation
         """
-        channel = f"{self.channel_prefix}response/matched-recommendations/"
-        response_dict = {"event": "match", "status": "success"}
+        channel = f"{self.response_channel}/matched-recommendations/"
+        response_data = {"event": "match", "status": "success"}
         data = json.loads(message.get("data"))
         recommendations = data.get("recommended_matches", [])
-        validated_records = {}
-        for record in recommendations:
-            market = self.markets_mapping.get(record.get("market_id"), None)
-            if market is None or market.readonly:
-                # The market is already finished or doesn't exist
-                continue
-
-            bid = offer_or_bid_from_json_string(json.dumps(record.get("bid")))
-            offer = offer_or_bid_from_json_string(json.dumps(record.get("offer")),
-                                                  record.get("bid").get("time"))
-
-            try:
-                validate_authentic_bid_offer_pair(
-                    bid,
-                    offer,
-                    record.get("trade_rate"),
-                    record.get("selected_energy")
-                    )
-
-                if not (offer.id in market.offers and bid.id in market.bids):
-                    # Offer or Bid either don't belong to market or were already matched
-                    raise InvalidBidOfferPair
-
-                if record.get("market_id") not in validated_records:
-                    validated_records[record.get("market_id")] = []
-
-                validated_records[record.get("market_id")].append(BidOfferMatch(
-                    bid,
-                    record.get("selected_energy"),
-                    offer,
-                    record.get("trade_rate")))
-            except InvalidBidOfferPair as ex:
-                # If validation fails or offer/bid were consumed
-                response_dict["status"] = "fail"
-                response_dict["message"] = "Validation Error"
-                logging.exception(f"Bid offer pair validation failed with error {ex}")
-                break
-        if response_dict["status"] == "success":
+        try:
+            validated_records = self._get_validated_bid_offer_match_list(recommendations)
             for market_id, records in validated_records.items():
                 market = self.markets_mapping.get(market_id)
-                if market.readonly:
-                    # The market has just finished
+                try:
+                    market.match_recommendation(records)
+                except (OfferNotFoundException, BidNotFoundException):
+                    # If the offer or bid have just been consumed
```

Comment:
When can this happen? It might be worth adding some more explanation to the comment. Also, we're losing any information about this scenario (since we're not logging the exception, but just continuing with the loop). This can be fine, but I'd just like to be sure that it's desired.

Reply:
This can happen in several cases, for example

1. an offer or bid can be deleted
2. myco could send the same offer or bid in several recommendations, one could be matched so the next time we encounter this offer this error will be raised

This is AFAIS, there could be other cases too, I'll update the comment accordingly